File: windows/app/OpenBurnBar.App.Presentation/Calendar/CalendarPageLayout.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace OpenBurnBar.App.Presentation.Calendar;

/// <summary>One card's persisted state: which card, whether shown, and how wide.</summary>
public sealed record CalendarCardConfig
{
    public CalendarCardConfig(CalendarCardKind kind, bool? isVisible = null, int? span = null)
    {
        Kind = kind;
        IsVisible = isVisible ?? true;
        Span = Math.Min(3, Math.Max(1, span ?? CalendarCardKindMetadata.DefaultSpan(kind)));
    }

    public CalendarCardKind Kind { get; }

    public bool IsVisible { get; set; }

    public int Span { get; set; }

    public string Id => CalendarCardKindMetadata.Id(Kind);
}

/// <summary>
/// The full panel layout: an ordered list of card configs, persisted as JSON.
/// Port of macOS <c>CalendarPageLayout</c> (<c>CalendarPageLayout.swift</c>) with
/// the same forward-compatible contract as <c>ChartsPageLayout</c> — unknown kinds
/// are dropped, missing kinds are appended with defaults, so upgrades never
/// clobber the user's arrangement. Pure + JSON-only so the persistence behavior
/// is unit-tested on the authoring host; the WinUI layer stores the encoded
/// string through the app's settings persistence.
/// </summary>
public sealed class CalendarPageLayout : IEquatable<CalendarPageLayout>
{
    private readonly List<CalendarCardConfig> _configs;

    public CalendarPageLayout(IEnumerable<CalendarCardConfig> configs)
    {
        _configs = Reconciled((configs ?? throw new ArgumentNullException(nameof(configs))).ToList());
    }

    /// <summary>Every registered card with default visibility/span, in registry order (fresh instance per call).</summary>
    public static CalendarPageLayout Default =>
        new(CalendarCardKindMetadata.All.Select(kind => new CalendarCardConfig(kind)));

    /// <summary>All card configs in display order (visible + hidden).</summary>
    public IReadOnlyList<CalendarCardConfig> Configs => _configs;

    /// <summary>Cards the grid actually renders, in order.</summary>
    public IReadOnlyList<CalendarCardConfig> VisibleConfigs =>
        _configs.Where(config => config.IsVisible).ToArray();

    /// <summary>Hidden cards, in order (drives the "Hidden Cards" restore menu).</summary>
    public IReadOnlyList<CalendarCardConfig> HiddenConfigs =>
        _configs.Where(config => !config.IsVisible).ToArray();

    // MARK: Mutations

    /// <summary>
    /// Moves <paramref name="kind"/> so it occupies the position currently held by
    /// <paramref name="target"/> (drag-to-reorder lands on a specific card, not an
    /// abstract index).
    /// </summary>
    public void Move(CalendarCardKind kind, CalendarCardKind target)
    {
        if (kind == target)
        {
            return;
        }

        int from = _configs.FindIndex(config => config.Kind == kind);
        int to = _configs.FindIndex(config => config.Kind == target);
        if (from < 0 || to < 0)
        {
            return;
        }

        CalendarCardConfig config = _configs[from];
        _configs.RemoveAt(from);
        _configs.Insert(to, config);
    }

    public void SetVisible(CalendarCardKind kind, bool visible)
    {
        CalendarCardConfig? config = _configs.FirstOrDefault(c => c.Kind == kind);
        if (config is not null)
        {
            config.IsVisible = visible;
        }
    }

    public void SetSpan(CalendarCardKind kind, int span)
    {
        CalendarCardConfig? config = _configs.FirstOrDefault(c => c.Kind == kind);
        if (config is not null)
        {
            config.Span = Math.Min(3, Math.Max(1, span));
        }
    }

    public void Reset()
    {
        _configs.Clear();
        _configs.AddRange(CalendarCardKindMetadata.All.Select(kind => new CalendarCardConfig(kind)));
    }

    // MARK: Row packing

    /// <summary>One packed row of the 3-column analytics grid.</summary>
    public sealed record Row(string Id, IReadOnlyList<CalendarCardConfig> Configs)
    {
        /// <summary>Columns this row occupies (≤ 3).</summary>
        public int SpanSum => Configs.Sum(config => Math.Min(3, Math.Max(1, config.Span)));
    }

    /// <summary>
    /// Packs configs greedily into rows up to 3 columns wide: a card joins the
    /// current row while the span sum stays ≤ 3, otherwise it starts a new row.
    /// Port of <c>CalendarAnalyticsPanel.rows(for:)</c>.
    /// </summary>
    public static IReadOnlyList<Row> PackRows(IEnumerable<CalendarCardConfig> configs)
    {
        var rows = new List<Row>();
        var current = new List<CalendarCardConfig>();
        int currentSpan = 0;
        foreach (CalendarCardConfig config in configs)
        {
            int span = Math.Min(3, Math.Max(1, config.Span));
            if (currentSpan + span > 3 && current.Count > 0)
            {
                rows.Add(new Row(string.Join("+", current.Select(c => c.Id)), current.ToArray()));
                current.Clear();
                currentSpan = 0;
            }

            current.Add(config);
            currentSpan += span;
        }

        if (current.Count > 0)
        {
            rows.Add(new Row(string.Join("+", current.Select(c => c.Id)), current.ToArray()));
        }

        return rows;
    }

    // MARK: Persistence

    private sealed class RawConfig
    {
        [JsonPropertyName("kind")]
        public string? Kind { get; set; }

        [JsonPropertyName("isVisible")]
        public bool? IsVisible { get; set; }

        [JsonPropertyName("span")]
        public int? Span { get; set; }
    }

    /// <summary>JSON round-trip tolerant of unknown kinds. See type comment.</summary>
    public static CalendarPageLayout Decode(string? json)
    {
        if (string.IsNullOrWhiteSpace(json))
        {
            return Default;
        }

        List<RawConfig>? raw;
        try
        {
            raw = JsonSerializer.Deserialize<List<RawConfig>>(json);
        }
        catch (JsonException)
        {
            return Default;
        }

        if (raw is null)
        {
            return Default;
        }

        var known = new List<CalendarCardConfig>();
        foreach (RawConfig entry in raw)
        {
            if (CalendarCardKindMetadata.KindForId(entry.Kind) is { } kind)
            {
                known.Add(new CalendarCardConfig(kind, entry.IsVisible, entry.Span));
            }
        }

        return new CalendarPageLayout(known);
    }

    /// <summary>Encodes the layout as a JSON array matching the macOS wire shape.</summary>
    public string Encode()
    {
        var raw = _configs.Select(config => new RawConfig
        {
            Kind = CalendarCardKindMetadata.Id(config.Kind),
            IsVisible = config.IsVisible,
            Span = config.Span,
        });
        return JsonSerializer.Serialize(raw);
    }

    /// <summary>Deduplicates and appends any kinds missing so every registered card has exactly one slot.</summary>
    private static List<CalendarCardConfig> Reconciled(List<CalendarCardConfig> configs)
    {
        var seen = new HashSet<CalendarCardKind>();
        var result = new List<CalendarCardConfig>();
        foreach (CalendarCardConfig config in configs)
        {
            if (seen.Add(config.Kind))
            {
                result.Add(config);
            }
        }

        foreach (CalendarCardKind kind in CalendarCardKindMetadata.All)
        {
            if (!seen.Contains(kind))
            {
                result.Add(new CalendarCardConfig(kind));
            }
        }

        return result;
    }

    public bool Equals(CalendarPageLayout? other) =>
        other is not null
        && _configs.Count == other._configs.Count
        && _configs.Zip(other._configs).All(pair =>
            pair.First.Kind == pair.Second.Kind
            && pair.First.IsVisible == pair.Second.IsVisible
            && pair.First.Span == pair.Second.Span);

    public override bool Equals(object? obj) => Equals(obj as CalendarPageLayout);

    public override int GetHashCode() => _configs.Count;
}

File: OpenBurnBarMobile/Views/Calendar/CalendarPageLayout.swift
import Foundation

// MARK: - Calendar Card Kind
//
// The registry of every analytics card the Calendar surface can render for
// the current day selection. Mirrors the macOS `CalendarCardKind` contract:
// each kind carries its own editorial metadata, and adding a card = add a
// case here, prepare its data in `CalendarSelectionSnapshot`, and give it a
// renderer arm in `CalendarAnalyticsPanel`.

enum CalendarCardKind: String, Codable, CaseIterable, Identifiable, Sendable {
    case kpis
    case burnOverSelection
    case providerMix
    case modelMix
    case hourOfDayHeatmap
    case projectFocus
    case cacheROI
    case reasoningShare

    var id: String { rawValue }

    var title: String {
        switch self {
        case .kpis: return "Key Numbers"
        case .burnOverSelection: return "Burn Over Selection"
        case .providerMix: return "Provider Mix"
        case .modelMix: return "Model Mix"
        case .hourOfDayHeatmap: return "When You Burn"
        case .projectFocus: return "Project Focus"
        case .cacheROI: return "Cache Savings"
        case .reasoningShare: return "Reasoning Share"
        }
    }

    /// One-line "why it matters" — rendered as the card's footer microcopy.
    var whyItMatters: String {
        switch self {
        case .kpis:
            return "The selection at a glance — spend, volume, sessions, cadence."
        case .burnOverSelection:
            return "Day-by-day spend across the selected days."
        case .providerMix:
            return "Where the money actually goes across providers."
        case .modelMix:
            return "Which models earn their keep — and which quietly dominate."
        case .hourOfDayHeatmap:
            return "Your true working rhythm inside the selection."
        case .projectFocus:
            return "Which projects the selected days actually funded."
        case .cacheROI:
            return "Prompt caching pays rent. This is the receipt."
        case .reasoningShare:
            return "Extended thinking is a hidden line item. Watch its share."
        }
    }

    var systemImage: String {
        switch self {
        case .kpis: return "number.square"
        case .burnOverSelection: return "chart.bar"
        case .providerMix: return "chart.pie"
        case .modelMix: return "cube.transparent"
        case .hourOfDayHeatmap: return "clock.badge"
        case .projectFocus: return "scope"
        case .cacheROI: return "archivebox"
        case .reasoningShare: return "brain"
        }
    }

    /// Grid span in columns on a 3-column grid (3 = full width). Surfaced in
    /// the UI as S/M/L sizes; compact widths clamp to the 2-column grid.
    var defaultSpan: Int {
        switch self {
        case .kpis, .burnOverSelection, .hourOfDayHeatmap:
            return 3
        default:
            return 1
        }
    }

    var defaultVisible: Bool { true }
}

// MARK: - Card configuration & page layout

/// One card's persisted state: which card, whether shown, and how wide.
struct CalendarCardConfig: Codable, Equatable, Identifiable, Sendable {
    var kind: CalendarCardKind
    var isVisible: Bool
    var span: Int

    var id: String { kind.rawValue }

    init(kind: CalendarCardKind, isVisible: Bool? = nil, span: Int? = nil) {
        self.kind = kind
        self.isVisible = isVisible ?? kind.defaultVisible
        self.span = min(3, max(1, span ?? kind.defaultSpan))
    }
}

/// The full panel layout: an ordered list of card configs, persisted as JSON
/// under `CalendarPageLayout.storageKey` (UserDefaults). Unknown kinds are
/// dropped, missing kinds are appended with defaults, so app upgrades never
/// clobber the user's arrangement.
struct CalendarPageLayout: Equatable, Sendable {
    static let storageKey = "calendarPageLayout.v1"

    private(set) var configs: [CalendarCardConfig]

    static var `default`: CalendarPageLayout {
        CalendarPageLayout(configs: CalendarCardKind.allCases.map { CalendarCardConfig(kind: $0) })
    }

    init(configs: [CalendarCardConfig]) {
        self.configs = Self.reconciled(configs)
    }

    /// Cards the grid actually renders, in order.
    var visibleConfigs: [CalendarCardConfig] {
        configs.filter(\.isVisible)
    }

    var hiddenConfigs: [CalendarCardConfig] {
        configs.filter { !$0.isVisible }
    }

    // MARK: Mutations

    /// Moves `kind` so it occupies the position currently held by `target`.
    mutating func move(_ kind: CalendarCardKind, toPositionOf target: CalendarCardKind) {
        guard kind != target,
              let from = configs.firstIndex(where: { $0.kind == kind }),
              let to = configs.firstIndex(where: { $0.kind == target }) else { return }
        let config = configs.remove(at: from)
        configs.insert(config, at: to)
    }

    /// Reorders configs to match a new kind order (edit-sheet `.onMove`).
    mutating func move(fromOffsets source: IndexSet, toOffset destination: Int) {
        configs.move(fromOffsets: source, toOffset: destination)
    }

    mutating func setVisible(_ kind: CalendarCardKind, _ visible: Bool) {
        guard let index = configs.firstIndex(where: { $0.kind == kind }) else { return }
        configs[index].isVisible = visible
    }

    mutating func setSpan(_ kind: CalendarCardKind, _ span: Int) {
        guard let index = configs.firstIndex(where: { $0.kind == kind }) else { return }
        configs[index].span = min(3, max(1, span))
    }

    mutating func reset() {
        self = .default
    }

    // MARK: Persistence

    /// JSON round-trip tolerant of unknown kinds. See type comment.
    static func decode(from data: Data) -> CalendarPageLayout {
        struct RawConfig: Codable {
            let kind: String
            let isVisible: Bool?
            let span: Int?
        }
        guard let raw = try? JSONDecoder().decode([RawConfig].self, from: data) else {
            return .default
        }
        let known = raw.compactMap { entry -> CalendarCardConfig? in
            guard let kind = CalendarCardKind(rawValue: entry.kind) else { return nil }
            return CalendarCardConfig(kind: kind, isVisible: entry.isVisible, span: entry.span)
        }
        return CalendarPageLayout(configs: known)
    }

    func encoded() -> Data? {
        try? JSONEncoder().encode(configs)
    }

    /// Deduplicates and appends any kinds missing from `configs` so every
    /// registered card always has exactly one slot.
    private static func reconciled(_ configs: [CalendarCardConfig]) -> [CalendarCardConfig] {
        var seen = Set<CalendarCardKind>()
        var result: [CalendarCardConfig] = []
        for config in configs where !seen.contains(config.kind) {
            seen.insert(config.kind)
            result.append(config)
        }
        for kind in CalendarCardKind.allCases where !seen.contains(kind) {
            result.append(CalendarCardConfig(kind: kind))
        }
        return result
    }

    // MARK: - Row packing

    /// Greedily packs the visible configs into rows for a `columns`-wide
    /// grid, preserving order: a card joins the current row when its span
    /// fits the remaining columns, otherwise it starts a new row. Spans are
    /// clamped to the grid width, so a span-3 card fills a 2-column compact
    /// grid. Pure and unit-tested (`CalendarPageLayoutTests`).
    static func packedRows(configs: [CalendarCardConfig], columns: Int) -> [[CalendarCardConfig]] {
        let columns = max(1, columns)
        var rows: [[CalendarCardConfig]] = []
        var currentRow: [CalendarCardConfig] = []
        var used = 0
        for config in configs {
            let span = min(columns, max(1, config.span))
            if used + span > columns, !currentRow.isEmpty {
                rows.append(currentRow)
                currentRow = []
                used = 0
            }
            currentRow.append(config)
            used += span
            if used >= columns {
                rows.append(currentRow)
                currentRow = []
                used = 0
            }
        }
        if !currentRow.isEmpty {
            rows.append(currentRow)
        }
        return rows
    }
}
